odxtools/basicstructure.py

```diff
@@ -122,52 +122,69 @@ def convert_physical_to_internal(self,
                                      triggering_coded_request: Optional[bytes],
                                      is_end_of_pdu: bool = True) -> bytes:
 
+        encode_state = EncodeState(
+            bytearray(),
+            parameter_values=cast(Dict[str, ParameterValue], param_value),
+            triggering_request=triggering_coded_request,
+            is_end_of_pdu=False)
+
         if not isinstance(param_value, dict):
-            raise EncodeError(
+            odxraise(
                 f"Expected a dictionary for the values of structure {self.short_name}, "
-                f"got {type(param_value)}")
+                f"got {type(param_value).__name__}", EncodeError)
+        elif encode_state.cursor_bit_position != 0:
+            odxraise(
+                f"Structures must be byte aligned, but "
+                f"{self.short_name} requested to be at bit position "
+                f"{encode_state.cursor_bit_position}", EncodeError)
+            encode_state.bit_position = 0
 
         # in strict mode, ensure that no values for unknown parameters are specified.
         if strict_mode:
-            param_names = [param.short_name for param in self.parameters]
-            for param_key in param_value:
-                if param_key not in param_names:
-                    odxraise(f"Value for unknown parameter '{param_key}' specified")
-
-        encode_state = EncodeState(
-            bytearray(),
-            dict(param_value),
-            triggering_request=triggering_coded_request,
-            is_end_of_pdu=False)
+            param_names = {param.short_name for param in self.parameters}
+            for param_value_name in param_value:
+                if param_value_name not in param_names:
+                    odxraise(f"Value for unknown parameter '{param_value_name}' specified "
+                             f"for structure {self.short_name}")
 
         for param in self.parameters:
-            if param == self.parameters[-1]:
-                # The last parameter is at the end of the PDU if the
-                # structure itself is at the end of the PDU. TODO:
-                # This assumes that the last parameter specified in
-                # the ODX is located last in the PDU...
+            if id(param) == id(self.parameters[-1]):
```

Comment:
Why?

Reply:
mainly  because comparing complex objects is potentially very expensive while comparing object IDs is cheap. (and at this particular place we get away with ids.)